File: plugin/src/main/java/qupath/gradle/QuPathGradleSettings.java
package qupath.gradle;

import org.gradle.api.GradleException;
import org.gradle.api.initialization.Settings;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.logging.Logger;
import org.gradle.api.logging.Logging;
import org.jetbrains.annotations.NotNull;

import java.net.URI;
import java.util.Set;

public class QuPathGradleSettings implements Plugin<Settings> {

    private static final Logger logger = Logging.getLogger(QuPathGradleSettings.class);

    @Override
    public void apply(@NotNull final Settings settings) {

        addSciJavaRepositories(settings);

        var extension = createExtension(settings);
        var gradle = settings.getGradle();

        gradle.settingsEvaluated(s -> {
            createCatalog(extension, s);
        });

        gradle.rootProject( p -> {
            applyToProject(extension, p);
        });

    }

    private QuPathExtension createExtension(Settings settings) {
        var extension = settings.getExtensions().create(
            "qupathExtension",
            QuPathExtension.class);
        // Default catalog name
        extension.getCatalogName().set("libs");
        return extension;
    }


    private void applyToProject(QuPathExtension extension, Project project) {
        logger.quiet("APPLYING TO PROJECT FROM JAVA");
        
        var props = project.getGradle().getExtensions().getExtraProperties();

        if (extension.getAutomaticModule().isPresent())
            props.set("extension.module", extension.getAutomaticModule().get());

        if (extension.getName().isPresent())
            props.set("extension.name", extension.getName().get());
        else
            throw new GradleException("QuPath extension name must be set");

        if (extension.getVersion().isPresent())
            props.set("extension.version", extension.getVersion().get());
        else
            throw new GradleException("QuPath extension version must be set (e.g. 0.1.0)");

        if (extension.getDescription().isPresent())
            props.set("extension.description", extension.getDescription().get());

        if (extension.getQupathVersion().isPresent())
            props.set("qupath.app.version", extension.getQupathVersion().get());
        else
            throw new GradleException("QuPath version must be set (e.g. 0.6.0)");
    }

    private void addSciJavaRepositories(Settings settings) {
        settings.getDependencyResolutionManagement().repositories(handler -> {
            logger.info("Adding SciJava Maven repositories");
            var urls = Set.of(
                    "https://maven.scijava.org/content/repositories/releases",
                    "https://maven.scijava.org/content/repositories/snapshots"
            );
            for (var url : urls) {
                handler.maven(repo -> repo.setUrl(URI.create(url)));
            }
        });
    }

    private void createCatalog(QuPathExtension extension, Settings settings) {
        String catalogName = extension.getCatalogName().getOrNull();
        if (catalogName == null) {
            logger.info("Version catalog not created (catalog name is null)");
            return;
        }
        var catalogBuilder = settings.getDependencyResolutionManagement().getVersionCatalogs();
        if (catalogBuilder.findByName(catalogName) != null) {
            logger.warn("QuPath version catalog named \"{}\" already exists - no new catalog will be created", catalogName);
            return;
        }
        catalogBuilder.create(catalogName, catalog -> {
            logger.info("Creating QuPath version catalog named \"{}\"", catalogName);
            catalog.from("io.github.qupath:qupath-catalog:" + extension.getQupathVersion().get());
        });
    }

}

File: plugin/src/main/kotlin/qupath/gradle/QuPathGradleSettings.kt
package qupath.gradle

import org.gradle.api.GradleException
import org.gradle.api.initialization.Settings
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.provider.Property
import java.net.URI

interface QuPathSettingsExtension {

    /**
     * The version of QuPath for which this extension is written.
     * This is used to determine which version catalog to use for dependencies.
     */
    val qupathVersion: Property<String>

    /**
     * The name of the extension, e.g. "qupath-extension-demo"
     */
    val name: Property<String>

    /**
     * The version of the extension, e.g. "0.1.0"
     */
    val version: Property<String>

    /**
     * A short description of the extension.
     */
    val description: Property<String>

    /**
     * The automatic module name for the extension, for insertion into the MANIFEST.MF file.
     */
    val automaticModule: Property<String>

    /**
     * The name of the version catalog to use for QuPath dependencies (default="libs")
     */
    val catalogName: Property<String>

}

class QuPathGradleSettings: Plugin<Settings> {

    override fun apply(settings: Settings) {

        addSciJavaRepositories(settings)

        val extension = createExtension(settings)

        settings.gradle.settingsEvaluated { s ->
            createCatalog(extension, s)
        }
        settings.gradle.rootProject { p ->
            applyToProject(extension, p)
        }

    }

    fun createExtension(settings: Settings): QuPathSettingsExtension {
        val extension = settings.extensions.create(
            "qupathExtension",
            QuPathSettingsExtension::class.java)
        // Default catalog name
        extension.catalogName.set("libs")
        return extension
    }


    fun applyToProject(extension: QuPathSettingsExtension, project: Project) {
        println("APPLYING TO PROJECT NOW")

        with (project.gradle.extensions.extraProperties) {
            if (extension.automaticModule.isPresent)
                set("extension.module", extension.automaticModule.get())

            if (extension.name.isPresent)
                set("extension.name", extension.name.get())
            else
                throw GradleException("QuPath extension name must be set")

            if (extension.version.isPresent)
                set("extension.version", extension.version.get())
            else
                throw GradleException("QuPath extension version must be set (e.g. 0.1.0)")

            if (extension.description.isPresent)
                set("extension.description", extension.description.get())

            if (extension.qupathVersion.isPresent)
                set("qupath.app.version", extension.qupathVersion.get())
            else
                throw GradleException("QuPath version must be set (e.g. 0.6.0)")
        }
    }

    fun addSciJavaRepositories(settings: Settings) {
        settings.dependencyResolutionManagement.repositories {
            println("Adding SciJava Maven repositories")
            it.maven {
                it.url = URI("https://maven.scijava.org/content/repositories/releases")
            }

            it.maven {
                it.url = URI("https://maven.scijava.org/content/repositories/snapshots")
            }
        }
    }

    fun createCatalog(extension: QuPathSettingsExtension, settings: Settings) {
        val catalogName: String = extension.catalogName.orNull ?: return
        with (settings.dependencyResolutionManagement.versionCatalogs) {
            if (names.contains(catalogName)) {
                println("QuPath version catalog named \"$catalogName\" already exists")
                return
            } else {
                create(catalogName) {
                    println("Creating QuPath version catalog named \"$catalogName\"")
                    it.from("io.github.qupath:qupath-catalog:${extension.qupathVersion.get()}")
                }
            }
        }
    }

}
